### ISSUE TITLE
[develop] Unable to install magento when document root is /pub

### ISSUE THREAD

**Issue by ihor-sviziev** [magento/magento2#10039]
### Preconditions
1. Install apache + php
2. Clone magento2 repository (develop branch)
3. Configure web server document root to magento2/pub directory.

### Steps to reproduce
1. Go to your magento2 URL without installation

### Expected result
1. You should see web setup wizard 

### Actual result

1. Too many redirects ![image](https://user-images.githubusercontent.com/1873745/27506855-6012fd5a-58cb-11e7-8005-aa7627898b39.png)
2. 
```
curl -I http://magento2.dev/
HTTP/1.1 302 Found
Server: nginx/1.13.0
Date: Sat, 24 Jun 2017 07:43:51 GMT
Content-Type: text/html; charset=UTF-8
Connection: keep-alive
X-Powered-By: PHP/7.0.20
Location: /setup/
X-Frame-Options: SAMEORIGIN
```

```
curl -I http://magento2.dev/setup/
HTTP/1.1 302 Found
Server: nginx/1.13.0
Date: Sat, 24 Jun 2017 07:43:58 GMT
Content-Type: text/html; charset=UTF-8
Connection: keep-alive
X-Powered-By: PHP/7.0.20
Location: /setup/
X-Frame-Options: SAMEORIGIN
```

We had similar issue in 2.1.x versions.

### Workaround
Change document root to magento2/ directory during setup.

**Comment by xtremeperf** [comment]
Perhaps I may not understand your question, but you are not supposed to be able to install Magento nor access the Web Setup Wizard from /pub.  I believe one of the primary reasons for serving your docroot from /pub is to specifically prevent web users having access to these sensitive areas of the system.

The webserver docroot should be set to /pub only after Magento is installed, configured and static files have been generated and populated in /pub.

Alternatively, you could create a subdomain (ex: install.domain.com) and configure your webserver's docroot to the Magento installed root folder.

If you are referring to _**upgrading**_ a previously installed copy of Magento, you should be able to access the web setup wizard from within the admin backend if you enable maintenance mode from the CLI first.

Regards,
Doug

**Comment by ihor-sviziev** [comment]
@xtremeperf This issue reproducing when you're trying to install new magento installation.

> The webserver docroot should be set to /pub only after Magento is installed, configured and static files have been generated and populated in /pub.

1. Where we can find this in docs?
2. In such case we should have error message "Magento is not installed, please execute \<link to documentation>following instructions</link to documentation>", but we're having redirect loop, that is not clear even for developers

**Comment by korostii** [comment]
Guys, please calm down, this is a known issue:
https://github.com/magento/magento2/issues/4159#issuecomment-232372120
https://github.com/magento/magento2/issues/5247#issuecomment-295247765

I've also submitted a ticket to the devdcos repository in order to have it marked as such:
https://github.com/magento/devdocs/issues/1120

**Comment by xtremeperf** [comment]
@korostii We are having a good conversion here. Good questions and nice helpful replys to each other.  Why do you want us to calm down?  I'm sorry if my replies are very direct and to the point, as that is simply the way my mind works.  I'm only trying to help out where I think that I can be of assistance.

@ihor-sviziev The Magento docs actually do not recommend setting your docroot to anything other than the install directory (/var/www/html/magento) at any time before or after installation.  Not that I have seen anyways.

http://devdocs.magento.com/guides/v2.0/install-gde/install-quick-ref.html

Setting your webserver's docroot to /magento/pub (the folder used to store only static generated public files) is a custom tactic used specifically for the purpose to restrict public access to vulnerable areas of the system.  Ask yourself this: do you want the public to be able to access your web setup wizard or any other directories outside of /pub?  If the answer is yes, then you should not have set the docroot to /pub in the first place.

As a developer, you can also inspect and compare the application, bootstrapping and entry point code located at magento/index.php and also at /magento/pub/index.php to see that the entry point code is different when you access the application from /pub.  Access to certain parts of the system are not available for security reasons.

I'm not sure the reason for "too many redirects". The answer probably lies within /magento/pub/.htaccess or similar.  I will take a look at the redirect code when I have a few minutes available.  If I can find the reason I will post it back here.

I hope these descriptions and explanations help clarify things.

Thanks,
Doug

**Comment by ihor-sviziev** [comment]
@xtremeperf actually I looked at the code, it's not .htaccess issue, in php we have check that app is installed, otherwise - 302 redirect to /setup/.

It's expected that some parts will not be available, but it shouldn't be in this way.

We had discussion in #1146 where I suggested to mark /index.php entrypoint as deprecated, but in order to do that - we need to fix issues related to using /pub as document root.

In latest commits in develop branch I saw that people already fixed issues with robots.txt when used /pub directory. I hope this issue will be fixed soon also.

@okorshenko could you look on this issue, can it be fixed by showing message like "Magento is not installed, please change your document root in server configuration to /magento2 directory and refresh the page to run setup wizard." instead of redirect loop?

**Comment by korostii** [comment]
@xtremeperf, I'm sorry for the bad choice of words, I intended no offense. What I meant is that the issue is long known by both the core team and the community, and it just might not be worth looking into on your own.

The comments I referenced show the core team saying that they tried to unravel this issue once and that they gave up on it shortly before the release of 2.1.
Which, in my opinion, probably means one of the following:
a) that this issue is too complex
b) this is, kind of, expected behavior and whatever you suggest to the core team (via a PR or otherwise) will get declined. The following quote seem to support this theory:
> This is an expected result if you use pub as your Magento docroot; it's designed this way so you cannot run the Web Setup Wizard in a production environment. Try setting the docroot to the Magento install directory and you should see the wizard again.

However, if that's case, the implementation could do better, obviously. Docs should probably explain this bug/feature/whatever-this-is explicitly, as well.

@ihor-sviziev,
> The Magento docs actually do not recommend setting your docroot to anything other than the install directory (/var/www/html/magento) at any time before or after installation

IMHO that's just messed up. From security perspective, it's quite obvious that the docroot on all production instances should be at `pub`.
